### PR TITLE
Utility bill calculation measure

### LINKED ISSUE #920: Utility bill calculation measure
SAM SDK?
- bundled in energyplus

Alternatives?
- dGen (Pieter Gagnon)
- python with openstudio (pycall)
- python OpenStudio measure that calls SAM?
- EMS?
- https://github.com/NREL/EnergyPlus/issues/9154

### PR DESCRIPTION
**Pull request by joseph-robertson**
## Pull Request Description

Addresses #920.

Creates a new `ReportUtilityBills` reporting measure that generates results_bills.csv (or results_bills.json), co-located with the EnergyPlus output. The file includes the following sections of output:

```
Total ($)

Electricity: Fixed ($)
Electricity: Marginal ($)
Electricity: PV Credit ($)
Electricity: Total ($)
 
Natural Gas: Fixed ($)
Natural Gas: Marginal ($)
Natural Gas: Total ($)

Fuel Oil: Total ($)

Propane: Total ($)

Wood Cord: Total ($)

Wood Pellets: Total ($)
 
Coal: Total ($)
```

## Checklist

Not all may apply:

- [ ] ~EPvalidator.xml has been updated~
- [x] Tests (and test files) have been updated
- [x] Documentation has been updated
- [x] Changelog has been updated
- [x] `openstudio tasks.rb update_measures` has been run
- [x] No unexpected regression test changes on CI (checked comparison artifacts)

### CODE REVIEW

File: docs/source/workflow_outputs.rst
Comment (joseph-robertson) [inline docs/source/workflow_outputs.rst:490]: @shorowit Should we populate these in `results_bills.csv` even when they are zero (i.e., when fuel type is not applicable)?

Comment (shorowit) [reply docs/source/workflow_outputs.rst:490]: Yes, let's populate them all regardless of fuel types in the building. That is most consistent with energy outputs (e.g., results_annual.csv) and eliminates surprises for users.

File: ReportUtilityBills/resources/Data/UtilityRates/Natural Gas.csv
Comment (shorowit) [inline ReportUtilityBills/resources/Data/UtilityRates/Natural Gas.csv:1]: Can we replace these fossil fuel CSVs with ones directly from EIA? For example, convert the second tab in the Download Series History link [here for natural gas](https://www.eia.gov/dnav/ng/ng_pri_sum_a_EPG0_PRS_DMcf_a.htm) to CSV and then use the latest available data (2021 or 2020, depending on state). Be sure to do unit conversions as needed.

**Review by shorowit — COMMENTED** [review]
Some initial comments/next steps.

File: workflow/run_simulation.rb
Comment (shorowit) [reply workflow/run_simulation.rb:67]: I am proposing updates to the default rates, so I think we can keep this idea.

File: ReportUtilityBills/resources/Data/UtilityRates/Electricity.csv
Comment (joseph-robertson) [reply ReportUtilityBills/resources/Data/UtilityRates/Electricity.csv:1]: But we are still not offering "auto" rates at this time, correct?

Comment (joseph-robertson) [reply ReportUtilityBills/resources/Data/UtilityRates/Electricity.csv:1]: Update: we are going to re-enable "auto" rates based on "latest data".

File: ReportUtilityBills/resources/Data/UtilityRates/Natural Gas.csv
Comment (joseph-robertson) [reply ReportUtilityBills/resources/Data/UtilityRates/Natural Gas.csv]: @shorowit Can you help me find rate CSVs for oil and propane?

Comment (shorowit) [reply ReportUtilityBills/resources/Data/UtilityRates/Natural Gas.csv]: Fuel oil: https://www.eia.gov/dnav/pet/xls/PET_PRI_WFR_A_EPD2F_PRS_DPGAL_W.xls
Propane: https://www.eia.gov/dnav/pet/xls/PET_PRI_WFR_A_EPLLPA_PRS_DPGAL_W.xls

Note that the values are weekly (and only for the half of the year when there's heating). Rather than just using the latest available week, which might be a temporary spike, I suggest we average over the last 26 values (i.e., an entire heating season).

For states that aren't explicitly called out, you'll have to use the region values instead. Here are the regions: https://www.eia.gov/todayinenergy/detail.php?id=4890

Comment (joseph-robertson) [reply ReportUtilityBills/resources/Data/UtilityRates/Natural Gas.csv]: Fuel oil: shows PADD 1 - 2
Propane: shows PADD 1 - 4

Comment (shorowit) [reply ReportUtilityBills/resources/Data/UtilityRates/Natural Gas.csv]: I guess we have to use the national value if no region is available. (Anytime we default to a region or national number for a given state, we should give a warning.)

**Comment by joseph-robertson** [comment]
Tests are failing, but let's review this PR before we update expected values. It's sort of tedious to update them.

**Comment by joseph-robertson** [comment]
@shorowit I want to wait for your review before updating expected test values. Your review might lead to rate changes for "auto".

**Comment by joseph-robertson** [comment]
> @shorowit I want to wait for your review before updating expected test values. Your review might lead to rate changes for "auto".

Didn't hear anything so I went ahead and generated expected values (using BEopt) based on rates corresponding to `Constants.Auto`.

File: docs/source/usage_instructions.rst
Comment (shorowit) [inline docs/source/usage_instructions.rst:50]: workflow/template.osw doesn't have any changes.

File: docs/source/workflow_outputs.rst
Comment (shorowit) [inline docs/source/workflow_outputs.rst:483]:
```suggestion
   Electricity: PV Credit ($)                     Annual production credit (negative value) for PV.
```

File: docs/source/workflow_outputs.rst
Comment (shorowit) [inline docs/source/workflow_outputs.rst:481]: Can we also add a simple "Total ($)", which is the sum of all the fuel totals, to the results_bills.csv and document that here?

File: workflow/template-report-utility-bills.osw
Comment (shorowit) [inline workflow/template-report-utility-bills.osw:1]: I don't think we need a separate OSW to show this. Just include the measure in template.osw.

File: workflow/tests/hpxml_translator_test.rb
Comment (shorowit) [inline workflow/tests/hpxml_translator_test.rb:233]: Can you also use `--add-utility-bills` in the `_run_xml` method so that it gets used for all workflow sample file runs? And make sure these utility bill results get captured in the CI artifacts/comparisons, either as new columns in results.csv or a new results_bills.csv (similar to the existing results_hvac_sizing.csv).

File: ReportUtilityBills/measure.rb
Comment (shorowit) [inline ReportUtilityBills/measure.rb:323]: DRY. 95% of this code is identical to code in the `energyPlusOutputRequests` section.

File: ReportUtilityBills/measure.rb
Comment (shorowit) [inline ReportUtilityBills/measure.rb:76]: Can we make all arguments optional instead of required? It's especially weird to require inputs for coal, propane, PV, etc. when a building might not have any of these things.

If an optional argument like `coal_marginal_rate` isn't provided but we see there is coal production in the home, I suggest we give a warning but continue to produce an output file (with zero for the coal $).

File: ReportUtilityBills/measure.rb
Comment (shorowit) [inline ReportUtilityBills/measure.rb:109]: These two can have `arg.setDefaultValue(Constants.Auto)`, right?

File: ReportUtilityBills/measure.rb
Comment (shorowit) [inline ReportUtilityBills/measure.rb:645]: 👍

File: ReportUtilityBills/measure.rb
Comment (shorowit) [inline ReportUtilityBills/measure.rb:781]: Lots of duplicate code with the ReportSimulationOutput measure. I suggest creating a new HPXMLtoOpenStudio/resources/output.rb where you consolidate code and reference from multiple reporting measures. You could also move the existing ReportSimulationOutput/constants.rb code into it and delete that constants.rb file.

File: workflow/run_simulation.rb
Comment (shorowit) [inline workflow/run_simulation.rb:87]: If all the measure arguments become optional, then these can all go away and we can use the measure defaults.

File: ReportUtilityBills/tests/utility_bills_test.rb
Comment (shorowit) [inline ReportUtilityBills/tests/utility_bills_test.rb:265]: Nice set of tests!

File: ReportUtilityBills/resources/Data/UtilityRates/NG_PRI_SUM_A_EPG0_PRS_DMCF_A.csv
Comment (shorowit) [inline ReportUtilityBills/resources/Data/UtilityRates/NG_PRI_SUM_A_EPG0_PRS_DMCF_A.csv:1]: Can you add a README.md that includes links to where these files came from?

**Review by shorowit — COMMENTED** [review]
This is coming along really well. Various comments below.

File: ReportUtilityBills/measure.rb
Comment (shorowit) [inline ReportUtilityBills/measure.rb:353]: Can you also `registerInfo` these values as we do [here](https://github.com/NREL/OpenStudio-HPXML/blob/d3a3bd73ba2c6a398fd722826b608ce5bbe20ba9/ReportSimulationOutput/measure.rb#L1191)? If you then use the `--debug` flag when calling the CLI, you can see the all registered values in the run.log. (It looks like we need this in the ReportHPXMLOutput measure too.)

Comment (shorowit) [reply ReportUtilityBills/measure.rb:353]: I manually added the registerInfo call to see what comes out:

```
Info: Registering 1305.57 for electricity_.
Info: Registering 223.92 for natural_gas_.
Info: Registering 0.0 for fuel_oil_.
Info: Registering 0.0 for propane_.
Info: Registering 0.0 for wood_cord_.
Info: Registering 0.0 for wood_pellets_.
Info: Registering 0.0 for coal_.
```

It looks like the "$" gets stripped, maybe use usd instead.

Also, it would be nice to register _every_ value that shows up in the output CSV/JSON, not just fuel totals.

File: docs/source/usage_instructions.rst
Comment (joseph-robertson) [reply docs/source/usage_instructions.rst:50]: workflow/template.osw now has changes.

File: workflow/run_simulation.rb
Comment (shorowit) [reply workflow/run_simulation.rb]: They vary significantly by region, but let's just use 0.015 $/kBtu as a default for all three.

File: workflow/template-report-utility-bills.osw
Comment (joseph-robertson) [reply workflow/template-report-utility-bills.osw]: Hmm. The issue with this is, e.g. `test_timeseries_daily_runperiod_Jan` of `output_report_test.rb`, doesn't run a full annual simulation which the utility bill measure requires...

Comment (joseph-robertson) [reply workflow/template-report-utility-bills.osw]: For this case, maybe we skip `ReportUtilityBills` instead of throw an error?

File: workflow/run_simulation.rb
Comment (shorowit) [inline workflow/run_simulation.rb:75]: Why are these here? Can't they be default argument values in ReportUtilityBills/measure.rb, as is done for every other argument?

File: workflow/template-report-utility-bills.osw
Comment (shorowit) [reply workflow/template-report-utility-bills.osw]: Maybe we can make the error-checking more nuanced. I suppose it's safe to do utility bill calculations as long as the electric rate is simple (not detailed) and there is no PV in the home. Would that help?

File: workflow/run_simulation.rb
Comment (joseph-robertson) [reply workflow/run_simulation.rb:75]: I guess it just didn't click for me that your previous comment was referring to _measure_ arguments.

File: ReportUtilityBills/measure.rb
Comment (shorowit) [inline ReportUtilityBills/measure.rb:199]:
```suggestion
        errors << 'A full annual simulation is required for calculating detailed utility bills.'
```

File: ReportUtilityBills/measure.rb
Comment (joseph-robertson) [reply ReportUtilityBills/measure.rb]: With defaults (see below), these rates will always be provided.

File: workflow/template-report-utility-bills.osw
Comment (joseph-robertson) [reply workflow/template-report-utility-bills.osw]: Yes.

File: workflow/tests/hpxml_translator_test.rb
Comment (joseph-robertson) [reply workflow/tests/hpxml_translator_test.rb:236]: Now adding `--add-utility-bills` for all sample xml files except for `base-location-capetown-zaf.xml` (no rates and so simulation errors out).

A `results_bills.csv` now gets exported to `workflow/tests/base_results` for all other xml files (excluding those with DSE).

File: ReportUtilityBills/measure.rb
Comment (joseph-robertson) [reply ReportUtilityBills/measure.rb]: I started down this route (create `output.rb`, move `constants.rb` into it, and start consolidating code). Can you help me fine tune this file?

File: workflow/tests/base_results/results_bills.csv
Comment (shorowit) [inline workflow/tests/base_results/results_bills.csv:1]: 👍 Great to have this file.

I assume it reflects the order of rows in the individual results_bills.csv files? If so, please put the total at the top of the file, not the bottom, like we do in results_annual.csv.

File: workflow/tests/base_results/results_bills.csv
Comment (shorowit) [inline workflow/tests/base_results/results_bills.csv:6]: It looks like the values in the individual results_bills.csv needs to be rounded to 2 decimals?

File: workflow/tests/base_results/results_bills.csv
Comment (shorowit) [inline workflow/tests/base_results/results_bills.csv:27]: I think this file isn't getting a result because it has DSE -- but it actually uses DSE of 1. Can you allow the utility bill calculation to occur if both heating & cooling DSEs are 1?

File: workflow/tests/hpxml_translator_test.rb
Comment (shorowit) [reply workflow/tests/hpxml_translator_test.rb:236]: Can we change it so that this measure never "errors out"? If we can't calculate bills, don't produce an output file, but return true to indicate that the measure/workflow was successful (like we do for DSE). I don't want users to not be able to get simulation results just because utility bills couldn't be calculated.

File: ReportUtilityBills/resources/util.rb
Comment (shorowit) [inline ReportUtilityBills/resources/util.rb:92]: Also for performance reasons, I moved the sum to outside the hourly loop. Between these two changes, the reporting measure runtime dropped from 1.0s to 0.15s.

**Review by shorowit — COMMENTED** [review]
Good work here, @joseph-robertson. This is really close.

File: ReportUtilityBills/measure.rb
Comment (shorowit) [inline ReportUtilityBills/measure.rb:662]: The fuel oil & propane code is 95% the same, can you combine then?

File: ReportUtilityBills/measure.rb
Comment (shorowit) [inline ReportUtilityBills/measure.rb:638]: Since we're no longer adjusting the rates provided by EIA, we're essentially treating the rate as an _average_, not _marginal_, rate and then we're adding fixed charges to it. So I'm realizing that we are essentially double counting the fixed charges. 

So we either need to A) not used the fixed charge when auto is provided or B) go back to doing some kind of adjustment. Having a hard time choosing between the two, but leaning towards the latter...

File: ReportUtilityBills/measure.rb
Comment (shorowit) [inline ReportUtilityBills/measure.rb:90]: I suggest we increase this to 12.0 just like electricity. [This source](https://www.aga.org/sites/default/files/aga_energy_analysis_-_natural_gas_utility_rate_structure.pdf) suggests the average was $11.25 in 2015.

File: ReportUtilityBills/tests/utility_bills_test.rb
Comment (joseph-robertson) [reply ReportUtilityBills/tests/utility_bills_test.rb:86]: Sample file `base-pv.xml` has two photovoltaic systems.

Comment (shorowit) [reply ReportUtilityBills/tests/utility_bills_test.rb:86]: Oh, I see. In that case, could you change these all to be e.g.:
```suggestion
    _load_timeseries(fuels, '../tests/PV_1kW.csv')
    @hpxml.pv_systems.each { |pv_system| pv_system.max_power_output = 1000.0 / @hpxml.pv_systems.size }
```

File: ReportUtilityBills/measure.rb
Comment (joseph-robertson) [reply ReportUtilityBills/measure.rb:643]: For B), here is where that adjustment was done previously: https://github.com/NREL/resstock/blob/v2.0.0/measures/UtilityBillCalculations/measure.rb#L378-L383.

Comment (shorowit) [reply ReportUtilityBills/measure.rb:643]: Thanks. Yes, let's go with that calculation again. I suggest creating a HouseholdConsumption.csv file that has Electricity and Natural Gas columns in it, and you can pull the data from the Electricity.csv and Natural gas.csv files in that measure.

**Review by shorowit — APPROVED** [review]
Looks good to me. Will merge once the CI comes back happy.

**Comment by shorowit** [comment]
Thanks @joseph-robertson!

**Pull request merged**
